shell/server/mail.js

```diff
@@ -535,6 +538,256 @@ HackSessionContextImpl.prototype.revokeApiToken = function (tokenId) {
+  this.server = server;
+  this.address = address;
+  this.port = port;
+}
+
+BoundUdpPortImpl.prototype.send = function(message, returnPort) {
+  this.server.send(message, 0, message.length, this.port, this.address);
+};
+
+IpInterfaceImpl.prototype.listenUdp = function (portNum, port) {
+  return new Promise(function (resolve, reject) {
+    var portMap = {};
+    var resolved = false;
+    var server = Dgram.createSocket('udp4'); // TODO(someday): handle ipv6 sockets too
+    server.bind(portNum);
+
+    server.on('listening', function () {
+      // Although UDP is connectionless, we don't resolve until here so that we can handle bind
+      // errors such as invalid host
+      resolved = true;
+      resolve({handle: server}); // server has a close method which is all we want from a handle
+    });
+
+    server.on('error', function (err) {
+      // TODO(someday): do something about errors after the promise is resolved
+      if (!resolved) {
+        reject(err);
+      } else {
+        console.error('error in listenUdp: ' + err);
+      }
+    });
+
+    var returnMap = {};
+    server.on('message', function (msg, rinfo) {
+      var address = rinfo.address + ']:' + rinfo.port;
+      var returnPort = returnMap[address];
+
+      if (!returnPort) {
+        returnMap[address] = returnPort = new BoundUdpPortImpl(server, rinfo.address, rinfo.port);
+      }
+      port.send(msg, returnPort);
+    });
+  });
+};
+
+HackSessionContextImpl.prototype.getIpInterface = function () {
+  return inMeteor((function () {
+    var grain = Grains.findOne(this.grainId, {fields: {userId: 1}});
+    var user = Meteor.users.findOne(grain.userId);
+
+    if (!user.isAdmin) {
+      throw new Error("This grain's owner does not have adequate permissions to call getIpInterface");
+    }
+
+    return {interface: new IpInterfaceImpl()};
+  }).bind(this));
+};
+
+var bits16 = Bignum(1).shiftLeft(16).sub(1);
+var bits32 = Bignum(1).shiftLeft(32).sub(1);
+
+var intToIpv4 = function (num) {
+  var part1 = num & 255;
+  var part2 = ((num >> 8) & 255);
+  var part3 = ((num >> 16) & 255);
+  var part4 = ((num >> 24) & 255);
+
+  return part4 + "." + part3 + "." + part2 + "." + part1;
+};
+
+var addressToString = function (address) {
+  var ipv6num = Bignum(address.upper64).shiftLeft(64).add(Bignum(address.lower64));
+
+  if (ipv6num.shiftRight(32).eq(bits16)) {
+    // this is an ipv4 address, we should return it as such
+    var ipv4num = ipv6num.and(bits32).toNumber();
+    return intToIpv4(ipv4num);
+  }
+  var hex = ipv6num.toString(16);
+  var numColons = 0;
+  var out = "";
+
+  for(var i = 0; i < hex.length; ++i) {
+    // start with lower bits of address and build the output in reverse
+    // this ensures that we can place a colon every 4 characters
+    out += hex[hex.length - 1 - i];
+    if ((i + 1) % 4 === 0) {
+      out += ':';
+      ++numColons;
+    }
+  }
+
+  // need to fill out colons on the upper end of the address til we have 7
```

Comment:
I think a double-colon can be used to mean "fill with as many zeroes as necessary" in an ipv6 address... though I might be making that up.